File: frontend/src/utils/classSorter.ts
/**
 * Natural ascending class sorter helper.
 * Correctly orders classes numerically by class number (1..12) and division (A..Z).
 * Example order:
 * Standard 1 - A, Standard 1 - B, Standard 2 - A ... Standard 9 - C, Standard 10 - A, Standard 10 - B
 */

export interface ClassSortKey {
  num: number;
  div: string;
  original: string;
}

export function extractClassSortKey(item: any): ClassSortKey {
  let nameStr = '';
  if (typeof item === 'string' || typeof item === 'number') {
    nameStr = String(item);
  } else if (item && typeof item === 'object') {
    if (item.class_name && item.division) {
      nameStr = `${item.class_name} ${item.division}`;
    } else if (item.class_name) {
      nameStr = item.class_name;
    } else if (item.name) {
      nameStr = item.name;
    } else if (item.label) {
      nameStr = item.label;
    } else {
      nameStr = String(item);
    }
  } else {
    nameStr = String(item || '');
  }

  // Extract numeric class number (e.g. 1, 2, 7, 8, 9, 10, 11, 12)
  const matchNum = nameStr.match(/\d+/);
  const num = matchNum ? parseInt(matchNum[0], 10) : 0;

  // Extract trailing division letter (e.g. A, B, C, D)
  const matchDiv = nameStr.match(/(?:-|\b(?:Division|Div|Sec|Section)\b)?\s*([A-Za-z])(?:\b|$)/i);
  const div = matchDiv ? matchDiv[1].toUpperCase() : '';

  return { num, div, original: nameStr };
}

export function sortClasses<T>(classesList: T[], keySelector?: (item: T) => any): T[] {
  if (!Array.isArray(classesList) || classesList.length === 0) return [];

  return [...classesList].sort((a, b) => {
    const itemA = keySelector ? keySelector(a) : a;
    const itemB = keySelector ? keySelector(b) : b;

    const keyA = extractClassSortKey(itemA);
    const keyB = extractClassSortKey(itemB);

    // 1. Primary sort: Class number (1, 2, 3 ... 9, 10, 11, 12)
    if (keyA.num !== keyB.num) {
      return keyA.num - keyB.num;
    }

    // 2. Secondary sort: Division (A, B, C...)
    if (keyA.div !== keyB.div) {
      return keyA.div.localeCompare(keyB.div);
    }

    // 3. Fallback: Full string comparison with numeric localeCompare
    return keyA.original.localeCompare(keyB.original, undefined, { numeric: true, sensitivity: 'base' });
  });
}

File: backend/app/core/class_sorter.py
import re
from typing import List, Any

def parse_class_key(c: Any):
    """
    Parses a class object, dict, or string to extract numeric class number and division
    for natural ascending sorting.
    Examples:
      'Standard 1 - A' -> (1, 'A', 'Standard 1 - A')
      'Standard 10 - B' -> (10, 'B', 'Standard 10 - B')
      'Class 2 Division C' -> (2, 'C', 'Class 2 Division C')
    """
    if hasattr(c, 'class_name'):
        div = getattr(c, 'division', '') or ''
        name_str = f"{c.class_name} {div}".strip()
    elif isinstance(c, dict):
        class_name = c.get('class_name') or c.get('name') or ''
        div = c.get('division') or ''
        name_str = f"{class_name} {div}".strip()
    else:
        name_str = str(c or '')

    # Extract class number (e.g. 1, 2, 7, 8, 9, 10, 11, 12)
    num_match = re.search(r'\d+', name_str)
    num = int(num_match.group()) if num_match else 0

    # Extract trailing division letter (e.g. A, B, C)
    div_match = re.search(r'(?:-|\b(?:Division|Div|Sec|Section)\b)?\s*([A-Za-z])(?:\b|$)', name_str)
    div = div_match.group(1).upper() if div_match else ''

    return (num, div, name_str)

def sort_classes_natural(classes_list: List[Any]) -> List[Any]:
    """
    Returns a new list sorted naturally by class number ascending (1..10) then division (A..Z).
    """
    if not classes_list:
        return []
    return sorted(classes_list, key=parse_class_key)
